### PR TITLE
mempool: fix nil pointer dereference

### LINKED ISSUE #5408: Nil pointer dereference in the mempool reactor
While running our usual batch of E2E tests on the just released v0.34-rc4, it seems that the newly introduced tx batching (commit ffe2742a6c47b88c67b32bee9dcf2b8b5b89b9ff part of #5321) seems to introduce a possible nil pointer dereference which causes the node to crash.

It seems to occurr in the `txs` function on line 275:
https://github.com/tendermint/tendermint/blob/7eb4e5c0b1de7aad987717d974ceec7012993f08/mempool/reactor.go#L274-L295

Did not investigate this much yet but on first glance it could be a concurrency issue -- at the end of the loop you have this check:
https://github.com/tendermint/tendermint/blob/7eb4e5c0b1de7aad987717d974ceec7012993f08/mempool/reactor.go#L291-L294

But each call to `Next()` seems to take a read lock:
https://github.com/tendermint/tendermint/blob/7eb4e5c0b1de7aad987717d974ceec7012993f08/libs/clist/clist.go#L115-L121

While you do check that `Next()` is non-nil, you invoke `Next()` again when assigning to `next`. And for this second call (I assume that the internal mutex is there because `e.next` can change concurrently) you don't check whether that is `nil`. I would suggest that you only call `Next()` once, check the value for nil and then set next to that value without calling `Next()` the second time.

Of course the root cause could also be something else as I didn't really investigate this further.

### PR DESCRIPTION
**Pull request by melekes**
previously, the second next could return nil, which would be the reason
for panic on line 275:

memTx := next.Value.(*mempoolTx)

Closes #5408

### CODE REVIEW

**Review by erikgrinaker — APPROVED** [review]
Makes sense. Have we verified that there aren't any other races here?

File: mempool/reactor.go [hunk 1]
```diff
@@ -288,10 +288,11 @@ func (memR *Reactor) txs(next *clist.CElement, peerID uint16, peerHeight int64)
 			batch = append(batch, memTx.tx)
 		}
 
-		if next.Next() == nil {
+		n := next.Next()
+		if n == nil {
 			return batch
 		}
-		next = next.Next()
+		next = n
```
Comment (marbar3778) [inline mempool/reactor.go:295]: should we add a regression test for this?

Comment (melekes) [reply mempool/reactor.go:295]: ideally 😄 kinda hard to test reactor at the moment

**Comment by melekes** [comment]
> Have we verified that there aren't any other races here?

none from the looks of it. I will try to write a test

File: mempool/reactor_test.go [hunk 2]
```diff
@@ -64,6 +65,66 @@ func TestReactorBroadcastTxsMessage(t *testing.T) {
 	waitForTxsOnReactors(t, txs, reactors)
 }
 
+// regression test for https://github.com/tendermint/tendermint/issues/5408
+func TestReactorConcurrency(t *testing.T) {
+	config := cfg.TestConfig()
+	const N = 2
```
Comment (erikgrinaker) [inline mempool/reactor_test.go:71]: What does this actually do? Seems like we're only testing two reactors regardless of this setting, in which case I think we can just drop this constant.

**Review by erikgrinaker — APPROVED** [review]

File: mempool/reactor_test.go (hunk 2 above)
Comment (melekes) [reply mempool/reactor_test.go:71]: we use it when creating reactors makeAndConnectReactors. the number can be increased to 4 for example. 2 is only required when we want to assert the order of transactions on the receiving side

Comment (erikgrinaker) [reply mempool/reactor_test.go:71]: Sure, but what purpose would it serve to start more reactors than the two we're testing?

Comment (melekes) [reply mempool/reactor_test.go:71]: I can't think of any

**Review by Karam19925 — COMMENTED** [review]
checked

**Pull request merged**